presto-main/src/main/java/io/prestosql/spi/expression/ConnectorExpressionUtil.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.expression;
+
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.Session;
+import io.prestosql.spi.type.Type;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.TypeProvider;
+import io.prestosql.sql.tree.AstVisitor;
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.ExpressionRewriter;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+import io.prestosql.sql.tree.NodeRef;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class ConnectorExpressionUtil
+{
+    private ConnectorExpressionUtil() {}
+
+    /**
+     * Produces {@link ConnectorExpression} translations for disjoint components in the {@param inputExpression} in a
+     * top-down manner. i.e. if an expression node is translatable, we do not consider its children.
+     */
+    public static Map<NodeRef<?>, ConnectorExpression> extractPartialTranslations(
+            Expression inputExpression,
+            Session session,
+            TypeAnalyzer typeAnalyzer,
+            TypeProvider typeProvider)
+    {
+        requireNonNull(inputExpression, "expressions is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(typeAnalyzer, "typeAnalyzer is null");
+        requireNonNull(typeProvider, "typeProvider is null");
+
+        ImmutableMap.Builder<NodeRef<?>, ConnectorExpression> partialTranslations = ImmutableMap.builder();
+        new PartialTranslator(typeAnalyzer.getTypes(session, typeProvider, inputExpression), partialTranslations).process(inputExpression);
+        return partialTranslations.build();
+    }
+
+    /**
+     * Traverse an expression recursively to replace subexpressions with an alternative expression, if specified in the
+     * {@param replacements} map.
+     */
+    public static Expression replaceNodeRefs(Expression expression, Map<NodeRef<?>, Expression> replacements)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new NodeRefReplacer(replacements), expression);
+    }
+
+    private static class NodeRefReplacer
```

Comment:
We have a very similar class for this already: `ReferenceAwareExpressionNodeInliner`